notebooks/ch11/gradient_stochastic.ipynb

```diff
@@ -0,0 +1,285 @@
+    "\n",
+    "Often, we calculate $\\nabla_{\\hat\\theta}L(\\hat\\theta, y)$ using the average gradient of the loss function $l(\\hat\\theta, y_i)$ across the entire dataset as a single batch. For this reason, this gradient update rule is often referred to as **batch gradient descent**.\n",
+    "\n",
+    "For example, the gradient of the MSE cost first requires us to find the gradient of the squared loss, $\\nabla_{\\hat\\theta} l(\\hat\\theta, y_i) = -2 (y_i - \\hat\\theta)$, for each of the $n$ observations in our dataset. We then compute the average of these results. This is shown below:\n",
+    "\n",
+    "$$\n",
+    "\\begin{align}\n",
+    "\\nabla_{\\hat\\theta} L(\\hat\\theta, y) &= -\\frac{2}{n} \\sum_{i=1}^{n}(y_i - \\hat\\theta) \\\\\n",
+    "&= \\frac{1}{n} \\sum_{i=1}^{n}-2(y_i - \\hat\\theta) \\\\\n",
+    "&= \\frac{1}{n} \\sum_{i=1}^{n} \\nabla_{\\hat\\theta} l(\\hat\\theta, y_i)\n",
+    "\\end{align}\n",
+    "$$\n",
+    "\n",
+    "\n",
+    "\n",
+    "This process is then repeated at each iteration $t$ of the gradient descent algorithm. For large $n$, this can become a computationally expensive problem to solve."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Stochastic Gradient Descent\n",
+    "\n",
+    "To circumvent the difficulty of computing a gradient across the entire training set, **stochastic gradient descent** approximates the overall gradient using a single randomly chosen data point. Since the observation is chosen randomly, we expect that using the gradient at each individual observation will eventually converge the algorithm to the same parameters as batch gradient descent. The general update formula for stochastic gradient descent is below, where $l(\\hat\\theta, y_i)$ is the loss function for a single data point:\n",
+    "\n",
+    "$$\n",
+    "\\hat\\theta_{t+1} = \\hat\\theta_t - \\alpha \\nabla_\\hat\\theta l(\\hat\\theta, y_i)\n",
+    "$$\n",
+    "\n",
+    "Returning back to our example using the MSE, we approximate the gradient of the mean squared error using the gradient of the squared loss of one data point. \n",
+    "\n",
+    "$$\n",
+    "\\begin{align}\n",
+    "\\nabla_{\\hat\\theta}L(\\hat\\theta, y) &\\approx \\nabla_{\\hat\\theta} l(\\hat\\theta, y_i) \\\\\n",
+    "&= -2(y_i - \\hat\\theta)\n",
+    "\\end{align}\n",
+    "$$\n",
+    "\n",
+    "Stochastic gradient descent relies on the random selection of individual observations. This is statistically founded, since the randomness means that $E[\\nabla_{\\hat\\theta}l(\\hat\\theta, y_i)] = \\nabla_{\\hat\\theta}L(\\hat\\theta, y)$. In practice, we choose these random data points by shuffling the training data and iteratively selecting from the shuffled data. An iteration through the shuffled data is called an **epoch**; at the end of every epoch, we re-shuffle the data to ensure that the model sees examples from various classes in an arbitrary order."
```

Comment:
I think it would be helpful to clarify that the end of an epoch is when all n training data points have been used for the weight update.

Also personally I found the "sees examples from various classes" description to be a little confusing. Perhaps we can reword it as "ensure that the model sees the data points in an arbitrary order."